stage1/init/common/mount.go

```diff
@@ -66,44 +92,84 @@ func GenerateMounts(ra *schema.RuntimeApp, volumes map[types.ACName]types.Volume
 
-			volumes[mp.Name] = emptyVol
-			ra.Mounts = append(ra.Mounts, schema.Mount{Volume: mp.Name, Path: mp.Path})
+			volumes[uniqName] = emptyVol
+			genMnts = append(genMnts,
+				mountWrapper{
+					mount: schema.Mount{
+						Volume: uniqName,
+						Path:   mp.Path,
+					},
+					dockerImplicit: dockerImplicit,
+				})
 		} else {
-			ra.Mounts = append(ra.Mounts, schema.Mount{Volume: vol.Name, Path: mp.Path})
+			genMnts = append(genMnts,
+				mountWrapper{
+					mount: schema.Mount{
+						Volume: vol.Name,
+						Path:   mp.Path,
+					},
+					dockerImplicit: false,
+				})
 		}
 	}
 
-	return ra.Mounts
+	return genMnts
 }
 
-// PrepareMountpoints creates and sets permissions for volumes.
-func PrepareMountpoints(path string, vol *types.Volume) error {
-	if vol.Kind == "empty" {
-		diag.Printf("creating an empty volume folder for sharing: %q", path)
-		err := os.MkdirAll(path, sharedVolPerm)
-		if err != nil {
-			return errwrap.Wrap(fmt.Errorf("could not create mount point for volume %q", vol.Name), err)
-		}
-		if err := os.Chown(path, *vol.UID, *vol.GID); err != nil {
-			return errwrap.Wrap(fmt.Errorf("could not change owner of %q", path), err)
-		}
-		mod, err := strconv.ParseUint(*vol.Mode, 8, 32)
-		if err != nil {
-			return errwrap.Wrap(fmt.Errorf("invalid mode %q for volume %q", *vol.Mode, vol.Name), err)
-		}
-		if err := os.Chmod(path, os.FileMode(mod)); err != nil {
-			return errwrap.Wrap(fmt.Errorf("could not change permissions of %q", path), err)
+// PrepareMountpoints creates and sets permissions for empty volumes.
+// If the mountpoint comes from a Docker image and it is an implicit empty
+// volume, we copy files from the image to the volume, see
+// https://docs.docker.com/engine/userguide/containers/dockervolumes/#data-volumes
+func PrepareMountpoints(volPath string, targetPath string, vol *types.Volume, dockerImplicit bool) error {
+	if vol.Kind != "empty" {
+		return nil
+	}
+
+	diag.Printf("creating an empty volume folder for sharing: %q", volPath)
+	if err := os.MkdirAll(volPath, 0770); err != nil {
+		return errwrap.Wrap(fmt.Errorf("error creating %q", volPath), err)
+	}
+	m, err := strconv.ParseUint(*vol.Mode, 8, 32)
+	if err != nil {
+		return errwrap.Wrap(fmt.Errorf("invalid mode %q for volume %q", *vol.Mode, vol.Name), err)
+	}
+	mode := os.FileMode(m)
+	Uid := *vol.UID
+	Gid := *vol.GID
+
+	if dockerImplicit {
+		fi, err := os.Stat(targetPath)
+		if err == nil {
+			// the directory exists in the image, let's set the same
+			// permissions and copy files from there to the empty volume
+			mode = fi.Mode()
+			Uid = int(fi.Sys().(*syscall.Stat_t).Uid)
+			Gid = int(fi.Sys().(*syscall.Stat_t).Gid)
+
+			if err := fileutil.CopyTree(targetPath, volPath, uid.NewBlankUidRange()); err != nil {
+				return errwrap.Wrap(fmt.Errorf("error copying image files to empty volume %q", volPath), err)
```

Comment:
"We regret to inform you that there was an error encountered copying some or all of the image files to the empty volume located at the path %q, and apologise for any inconvenience this fault may cause you"